File: RxTango/java/examples/AlarmMonitor.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 11+
//REPOS mavencentral,jtango=https://maven.pkg.github.com/scientific-software-hub/JTango
//DEPS io.reactivex.rxjava3:rxjava:3.1.8
//DEPS org.reactivestreams:reactive-streams:1.0.4
//DEPS org.waltz.tango:ez:1.6.3
//DEPS org.waltz.tango.orb:tangorb:1.6.3
//SOURCES ../src/RxTango.java ../src/RxTangoAttribute.java

import io.reactivex.rxjava3.core.Flowable;
import org.tango.client.rx.RxTangoAttribute;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Fan-in alarm monitor: polls N attributes from N (possibly different) devices,
 * merges all streams into one, and emits an alarm line whenever any value
 * exceeds the threshold.
 *
 * Each device is polled independently. A failure on one device does not affect
 * the others — the stream simply logs the error and continues.
 *
 * This is the fan-in pattern: N independent sources → 1 unified alarm stream.
 *
 * Usage:
 *   jbang AlarmMonitor.java <threshold> <interval-ms> <device> <attr> [<device> <attr> ...]
 *
 * Example (alert when |value| > 200, poll every 500 ms):
 *   jbang AlarmMonitor.java 200 500 \
 *       tango://localhost:10000/sys/tg_test/1 double_scalar \
 *       tango://localhost:10000/sys/tg_test/1 long_scalar
 */
public class AlarmMonitor {

    record Target(String device, String attribute) {}

    public static void main(String[] args) throws Exception {
        if (args.length < 4) {
            System.err.println("Usage: AlarmMonitor <threshold> <interval-ms> <device> <attr> [<device> <attr> ...]");
            System.exit(1);
        }

        double threshold = Double.parseDouble(args[0]);
        long   intervalMs = Long.parseLong(args[1]);

        List<Target> targets = new ArrayList<>();
        for (int i = 2; i + 1 < args.length; i += 2)
            targets.add(new Target(args[i], args[i + 1]));

        if (targets.isEmpty()) {
            System.err.println("No device/attribute pairs found.");
            System.exit(1);
        }

        System.out.printf("Monitoring %d source(s), threshold ±%.2f, poll every %d ms — Ctrl+C to stop%n",
                targets.size(), threshold, intervalMs);

        // Build one polling stream per target, then merge them all into a single alarm stream.
        // Each stream is independent: an error on one device doesn't kill the others.
        List<Flowable<String>> streams = new ArrayList<>();
        for (Target t : targets) {
            Flowable<String> stream = Flowable.interval(intervalMs, TimeUnit.MILLISECONDS)
                    .flatMapSingle(tick ->
                            Flowable.fromPublisher(new RxTangoAttribute<>(t.device(), t.attribute()))
                                    .firstOrError()
                                    .map(v -> ((Number) v).doubleValue())
                    )
                    .filter(v -> Math.abs(v) > threshold)
                    .map(v -> String.format("ALARM  [%s]  %s/%s = %.4f  (threshold ±%.2f)",
                            Instant.now(), t.attribute(), t.device(), v, threshold))
                    .onErrorResumeNext(err -> {
                        System.err.printf("Stream error for %s/%s: %s — continuing%n",
                                t.device(), t.attribute(), err.getMessage());
                        return Flowable.empty();
                    });
            streams.add(stream);
        }

        Flowable.merge(streams)
                .blockingSubscribe(
                        System.out::println,
                        err -> System.err.println("Fatal: " + err.getMessage())
                );
    }
}

File: RxTango/cpp/examples/alarm_monitor.cpp
/**
 * Alarm fan-in — watch N devices, fire when any exceeds a threshold.
 *
 * One poll stream per device; merge() fans them into a single alarm stream;
 * on_error_resume_next() isolates per-device failures.
 *
 * Mirrors Python's alarm_monitor.py and Java's AlarmMonitor.java.
 *
 * Usage:
 *   ./alarm_monitor [threshold] [interval-ms] [device1] [device2] ...
 *   defaults: threshold=0.5  interval=500  (two reads from sys/tg_test/1)
 */

#include <chrono>
#include <csignal>
#include <iostream>
#include <string>
#include <vector>

#include <rxcpp/rx.hpp>
#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const double threshold   = argc > 1 ? std::stod(argv[1]) : 0.5;
    const int    interval_ms = argc > 2 ? std::stoi(argv[2]) : 500;

    std::vector<std::string> devices;
    for (int i = 3; i < argc; ++i) devices.emplace_back(argv[i]);
    if (devices.empty())
        devices = { "tango://localhost:10000/sys/tg_test/1",
                    "tango://localhost:10000/sys/tg_test/1" };

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Alarm monitor — threshold = " << threshold << "  (Ctrl+C to stop)\n\n";

    // Build one poll stream per device, merge all into one alarm stream
    std::vector<rxcpp::observable<std::pair<std::string,double>>> streams;
    for (auto& dev : devices) {
        streams.push_back(
            rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
                .flat_map([dev](long) {
                    return rxtango::read_attribute<double>(dev, "double_scalar")
                        .map([dev](double v) { return std::make_pair(dev, v); })
                        .on_error_resume_next([dev](std::exception_ptr) {
                            // isolate per-device failure; return sentinel pair
                            return rxcpp::observable<>::just(
                                std::make_pair(dev, std::numeric_limits<double>::quiet_NaN()));
                        });
                })
        );
    }

    // Merge all streams and filter for alarms
    rxcpp::observable<>::iterate(streams)
        .flat_map([](rxcpp::observable<std::pair<std::string,double>> s) { return s; })
        .filter([threshold](std::pair<std::string,double> p) {
            return !std::isnan(p.second) && std::abs(p.second) > threshold;
        })
        .subscribe(
            [](std::pair<std::string,double> p) {
                std::cout << "  ALARM  " << p.first << " = " << p.second << "\n";
            },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    return 0;
}
